nexus/db-queries/src/db/datastore/ereport.rs

```diff
@@ -0,0 +1,423 @@
+        pagparams: &DataPageParams<'_, DbEna>,
+    ) -> ListResultVec<SpEreport> {
+        opctx.authorize(authz::Action::ListChildren, &authz::FLEET).await?;
+        paginated(sp_dsl::sp_ereport, sp_dsl::ena, pagparams)
+            .filter(sp_dsl::restart_id.eq(restart_id.into_untyped_uuid()))
+            .filter(sp_dsl::time_deleted.is_null())
+            .select(SpEreport::as_select())
+            .load_async(&*self.pool_connection_authorized(opctx).await?)
+            .await
+            .map_err(|e| public_error_from_diesel(e, ErrorHandler::Server))
+    }
+
+    /// List unique ereporter restarts for the given serial number.
+    pub async fn ereporter_restart_list_by_serial(
+        &self,
+        opctx: &OpContext,
+        serial: String,
+    ) -> ListResultVec<EreporterRestartBySerial> {
+        use diesel::dsl::{count_distinct, min};
+        opctx.authorize(authz::Action::ListChildren, &authz::FLEET).await?;
+
+        let conn = &*self.pool_connection_authorized(opctx).await?;
+        let sp_rows = sp_dsl::sp_ereport
+            .filter(
+                sp_dsl::serial_number
+                    .eq(serial.clone())
+                    .and(sp_dsl::time_deleted.is_null()),
+            )
+            .group_by((sp_dsl::restart_id, sp_dsl::sp_slot, sp_dsl::sp_type))
+            .select((
+                sp_dsl::restart_id,
+                sp_dsl::sp_type,
+                sp_dsl::sp_slot,
+                min(sp_dsl::time_collected),
+                count_distinct(sp_dsl::ena),
+            ))
+            .order_by(sp_dsl::restart_id)
+            .load_async::<(Uuid, SpType, SqlU16, Option<DateTime<Utc>>, SqlU32)>(
+                conn,
+            )
+            .await
+            .map_err(|e| public_error_from_diesel(e, ErrorHandler::Server))?;
+        let host_os_rows = host_dsl::host_ereport
+            .filter(
+                host_dsl::sled_serial
+                    .eq(serial)
+                    .and(host_dsl::time_deleted.is_null()),
+            )
+            .group_by((host_dsl::restart_id, host_dsl::sled_id))
+            .select((
+                host_dsl::restart_id,
+                host_dsl::sled_id,
+                min(host_dsl::time_collected),
+                count_distinct(host_dsl::ena),
+            ))
+            .order_by(host_dsl::restart_id)
+            .load_async::<(Uuid, Uuid, Option<DateTime<Utc>>, SqlU32)>(conn)
+            .await
+            .map_err(|e| {
+                public_error_from_diesel(e, ErrorHandler::Server)
+                    .internal_context("listing SP ereports")
+            })?;
+
+        let sp_reporters = sp_rows.into_iter().filter_map(|(restart_id, sp_type, sp_slot, first_seen, ereports)| {
+            let ereports = ereports.into();
+            let first_seen_at = match first_seen {
+                Some(t) => t,
+                None => {
+                    debug_assert!(false, "ELIZA: why would min(time_collected) be none?");
```

Comment:
This and the `debug_assert`  below probably shouldn't be here?

Reply:
After consulting the docs, the only reasons the `min()` function would return `NULL` are:
1. there are no rows in the result set
2. all values for the `min`'d column are `NULL`

In this case, because this is a `GROUP BY` query, the "no rows returned" case wouldn't matter, since the `SELECT` would just not return anything, rather than returning a `NULL` min value. And, the `min`'d column has a `NOT NULL` constraint, so we shouldn't ever hit the "all values are `NULL`" either. It's probably fine to just unconditionally `expect` this...

Reply:
Okay, based on that, this will now unconditionally panic, because I think we should never see a `NULL` here: https://github.com/oxidecomputer/omicron/pull/8296/commits/372d582602d61d7fa98e80eee99a5f7a937d9bda